src/openrct2-ui/windows/Research.cpp

```diff
@@ -154,117 +113,218 @@ static constexpr const StringId ResearchStageNames[] = {
-    w->event_handlers = window_research_page_events[0];
-    w->pressed_widgets = 0;
-    w->disabled_widgets = 0;
-    WindowInitScrollWidgets(*w);
+        hold_down_widgets = 0;
+        widgets = window_research_page_widgets[p];
+        disabled_widgets = 0;
+        pressed_widgets = 0;
 
-    return w;
-}
+        Invalidate();
+        if (p == WINDOW_RESEARCH_PAGE_DEVELOPMENT)
+        {
+            width = WW_DEVELOPMENT;
+            height = WH_DEVELOPMENT;
+        }
+        else
+        {
+            width = WW_FUNDING;
+            height = WH_FUNDING;
+        }
+        WindowEventResizeCall(this);
+        WindowEventInvalidateCall(this);
 
-#pragma region Development page
+        WindowInitScrollWidgets(*this);
+        Invalidate();
+    }
 
-/**
- *
- *  rct2: 0x006B6B38
- */
-static void WindowResearchDevelopmentMouseup(WindowBase* w, WidgetIndex widgetIndex)
-{
-    switch (widgetIndex)
+    void OnUpdate() override
     {
-        case WIDX_CLOSE:
-            WindowClose(*w);
-            break;
-        case WIDX_TAB_1:
-        case WIDX_TAB_2:
-            WindowResearchSetPage(w, widgetIndex - WIDX_TAB_1);
-            break;
-        case WIDX_LAST_DEVELOPMENT_BUTTON:
-            News::OpenSubject(News::ItemType::Research, gResearchLastItem->rawValue);
-            break;
+        // Tab animation
+        if (++frame_no >= window_research_tab_animation_loops[page])
+            frame_no = 0;
+        InvalidateWidget(WIDX_TAB_1 + page);
     }
-}
 
-/**
- *
- *  rct2: 0x0069CBA6
- */
-static void WindowResearchDevelopmentUpdate(WindowBase* w)
+    void OnMouseDown(WidgetIndex widgetIndex) override
+    {
+        if (page == WINDOW_RESEARCH_PAGE_FUNDING)
+        {
+            WindowResearchFundingMouseDown(this, widgetIndex, WIDX_RESEARCH_FUNDING);
+        }
+    }
+
+    void OnMouseUp(WidgetIndex widgetIndex) override
+    {
+        // Switch tab or close
+        switch (widgetIndex)
+        {
+            case WIDX_CLOSE:
+            {
+                Close();
+                break;
+            }
+            case WIDX_TAB_1:
+            case WIDX_TAB_2:
+            {
+                SetPage(widgetIndex - WIDX_TAB_1);
+                break;
+            }
+        }
+
+        // Process mouse up for specific tab
+        switch (page)
+        {
+            case WINDOW_RESEARCH_PAGE_DEVELOPMENT:
+            {
+                WindowResearchDevelopmentMouseUp(widgetIndex, WIDX_CURRENTLY_IN_DEVELOPMENT_GROUP);
+                break;
+            }
+            case WINDOW_RESEARCH_PAGE_FUNDING:
+            {
+                WindowResearchFundingMouseUp(widgetIndex, WIDX_RESEARCH_FUNDING);
+                break;
+            }
+        }
+    }
+
+    void OnDropdown(WidgetIndex widgetIndex, int32_t selectedIndex) override
+    {
+        if (page == WINDOW_RESEARCH_PAGE_FUNDING)
+        {
+            WindowResearchFundingDropdown(widgetIndex, selectedIndex, WIDX_RESEARCH_FUNDING);
```

Comment:
Sounds like `WIDX_RESEARCH_FUNDING` can be built into the function, since all callers use the same parameter.

Reply:
This is also called from the Finance window (Line 288). The enum values are different in each window. This also ties into the comment on this:
```
    // Offset the widget index to allow reuse from other windows
    baseWidgetIndex = baseWidgetIndex - WIDX_RESEARCH_FUNDING;
```
Windows that resuse these methods will have a different value for WIDX_RESEARCH_FUNDING. So the WIDX components need to be referenced relative to whatever index they start at. For example in the Finance window WIDX_TRANSPORT_RIDES = 14, whereas in the Research window WIDX_TRANSPORT_RIDES = 10. 

Not sure if I have explained that well? This offset logic already existed in this class, I have just expanded its use.

Reply:
I think it's a good enough explanation, maybe we should just extract the logic to a static function on the Research file with a lengthier comment, that would also prevent the comment and the code duplication everywhere